### PR TITLE
Add scattering for ModuleReplacementAnnotation

### PR DESCRIPTION
**Pull request by albertchen-sifive**
Scatters a `ModuleReplacementAnnotation`.

```json5
{
  "class": "sifive.enterprise.grandcentral.ModuleReplacementAnnotation",

  // the list of intances/modules to replace
  "targets": [
    "~Top|Top/child:Child",
    "~Top|Top/childWrapper:ChildWrapper/child:Child"
  ],

  // the replacement module
  "circuit": "",

  // annotations associated with the replacement module
  "annotations": [],

  // prefix and output sub-directory of the replacement module
  "circuitPackage": "other",

  // dontTouch all the ports of replaced modules
  "dontTouches":[
    "~Top|Child>in",
    "~Top|Child>out"
  ]
}
```

This adds a parent annotation to the circuit, annotates each member of `targets` with an id and the parent id, and adds dontTouches to each entry in `dontTouches`. Mostly copied from `SignalDriverAnnotation`

### CODE REVIEW

File: lib/Dialect/FIRRTL/Import/FIRAnnotations.cpp
Comment (seldridge) [inline lib/Dialect/FIRRTL/Import/FIRAnnotations.cpp:1732]: As I mentioned to you offline, the repeated use of this canonicalize + non-local expansion would be great as a utility method (in a follow-on PR/commit).

File: lib/Dialect/FIRRTL/Import/FIRAnnotations.cpp
Comment (seldridge) [inline lib/Dialect/FIRRTL/Import/FIRAnnotations.cpp:1746]: I have not audited the scattering code to guarantee that the IDs are unique across all scattered annotations.  However, if they are, then the "class" can be dropped.  (I realize it doesn't work this way in other scatterings.)

File: lib/Dialect/FIRRTL/Import/FIRAnnotations.cpp
Comment (seldridge) [inline lib/Dialect/FIRRTL/Import/FIRAnnotations.cpp:1750]: This indentation looks a bit weird.  Is this what `clang-format` is doing here?  You may need to combine the strings and then let `clang-format` split them.

File: test/Dialect/FIRRTL/annotations.fir
Comment (seldridge) [inline test/Dialect/FIRRTL/annotations.fir:1118]: Mega nit:
```suggestion
; Grand Central's ModuleReplacementAnnotation is properly scattered to the circuit
; and the targeted operations.

circuit Top : %[[{
```

File: test/Dialect/FIRRTL/annotations.fir
Comment (seldridge) [inline test/Dialect/FIRRTL/annotations.fir:1125]: It would be good to populate these for this test and make sure that they are copied over correctly.

File: test/Dialect/FIRRTL/annotations.fir
Comment (seldridge) [inline test/Dialect/FIRRTL/annotations.fir:1152]: Nit: the `@Child...` can be removed in the check.  Generally, it's good practice to keep the tests as terse as possible and not check more than is needed.  E.g., this doesn't really need to check the ports.

File: lib/Dialect/FIRRTL/Import/FIRAnnotations.cpp
Comment (seldridge) [inline lib/Dialect/FIRRTL/Import/FIRAnnotations.cpp:1715]: I was surprised that this did _not_ include a `targets` field in the circuit-level annotation.  However, all the information in there is already handled based on the ordering of the target IDs.

This is fine, but will miss a situation where the _last_ target was deleted... e.g., if the targets are `[4, 5, 6]` and `6` gets deleted by a pass, then this would be lost. This can either be fixed by including a `numTargets` field here or by just doing a `targets = [4, 5, 6]` array here.

What do you think?

**Review by seldridge — COMMENTED** [review]
Looks awesome, @albertchen-sifive! Thank you for your first CIRCT contribution!

I have one question about tracking deleted targets.  Everything else looks great.

File: lib/Dialect/FIRRTL/Import/FIRAnnotations.cpp
Comment (albertchen-sifive) [reply lib/Dialect/FIRRTL/Import/FIRAnnotations.cpp:1715]: I don't think module replacement annotation needs to keep track of deleted targets? in SFC we don't really care about which ones get deleted, the update method just silently removes targets from the list if they get deleted.

Comment (seldridge) [reply lib/Dialect/FIRRTL/Import/FIRAnnotations.cpp:1715]: If silently dropping deleted modules/instances is the correct behavior, then not recording the targets in the circuit-level annotation is the most efficient solution. 💯

**Review by seldridge — APPROVED** [review]
This just needs a `clang-format` and it should be good to go.

File: lib/Dialect/FIRRTL/Import/FIRAnnotations.cpp
Comment (albertchen-sifive) [reply lib/Dialect/FIRRTL/Import/FIRAnnotations.cpp:1746]: we can also get rid of the `targetId` field right? I don't think we need to preserve order of the `targets` array. In SFC when we emit config files we sort by the module/instance path string instead of preserving the original ordering.

Comment (seldridge) [reply lib/Dialect/FIRRTL/Import/FIRAnnotations.cpp:1746]: Yeah, I don't see a need to preserve order. 👍

**Review by seldridge — APPROVED** [review]

**Comment by seldridge** [comment]
It looks like the test needs to be updated to strip the now removed `targetId` field. 😅

**Comment by seldridge** [comment]
Thanks for the patch, @albertchen-sifive! I'll merge this once it passes regressions. If you'd like commit access (to merge your PRs yourself), please read the [LLVM Developer Policy](https://llvm.org/docs/DeveloperPolicy.html) for instructions on how to get that.

**Comment by seldridge** [comment]
*One final thing: `clang-tidy` is unhappy about variables that start with capital letters. You can change `NLATargets` to `nlaTargets` to align with the `clang-tidy` policies.

**Pull request merged**
